### ISSUE TITLE
[Core] Dangling actor handle held at Ray client

### ISSUE THREAD

**Issue by jianoaix** [ray-project/ray#25237]
TL;DR user tried to use an actor at Ray client (via the actor handle it kept) but the actor was already gone in Ray cluster, resulting in KeyError.

**What happened**
The user was running a series of tune trials, and they restarted Ray cluster and reconnected for each trial from Ray client.
The first trial can work fine but in the second trial it crashed on KeyError, when it tried to load data via ray.data.read_parquet().

**Hypothesis**
When we create Dataset we use a stats actor to keep execution stats. We reuse the same stats actor for all Datasets created by the same process. This is done via keeping the actor handle in a global variable after the first time the actor was created.

The issue is the process (Ray client) doesn't reset or restart when it restarts the Ray cluster in each trial. So after the first trial and the Ray cluster restarted, the actor was already gone in the cluster. The actor handle kept in Ray client process is dangling.

**Verification**
First, we verified that the actor handle was indeed the same as the one used in the first trial.
Second, we verified that it will work, if we manually nulled out the stats actor handle and re-created it in each trial.

**Solutions**
- Option 1: keep a node_id together with the actor handle, and reuse stats actor only if Datasets are created under same node_id
- Option 2: use the client_id, which is uuid per connection, to detect the change and reuse the stats actor only if Datasets are created under same client_id

**Comment by jianoaix** [comment]
@iycheng @ckw017

**Comment by ckw017** [comment]
Ah sorry, just to make sure we're on the same page are you asking for a change in core/client here?

**Comment by jianoaix** [comment]
@ckw017 Yes. I think the lifetime of the references kept at Ray client should not be tied to the lifetime of client process, because this process can outlive the Ray cluster. So a better lifetime scope would be per connection, i.e. if it is disconnected or get reconnected, those references (actor handles, object refs) should be invalidated/cleared.

**Comment by ckw017** [comment]
What do you mean by invalidated/cleared in this case? In some sense they're already invalidated (using them will fail). Do you want the failure to happen on the client side instead of on the server?

**Comment by jianoaix** [comment]
The desired outcome, if we fix it, is that there is not failure on either side at all. By invalidated/cleared I mean we make "some_ref = None", rather than still have "some_ref = Actor(SomeActor, 02549a3c3f0b1e6b34f1d50c01000000)".
